### PR TITLE
KAFKA-13957: Fix flaky shouldQuerySpecificActivePartitionStores test

### PR DESCRIPTION
**Pull request by mdedetrich**
Currently the tests fail because there is a missing predicate in the `retrievableException` which causes the test to fail, i.e. the current predicates

```java
containsString("Cannot get state store source-table because the stream thread is PARTITIONS_ASSIGNED, not RUNNING"),
containsString("The state store, source-table, may have migrated to another instance"),
containsString("Cannot get state store source-table because the stream thread is STARTING, not RUNNING")
```
wasn't complete. Another one needed to be added, namely "The specified partition 1 for store source-table does not exist.". This is because its possible for 

```java
assertThat(getStore(kafkaStreams2, storeQueryParam2).get(key), is(nullValue()));
```         
or

```java
assertThat(getStore(kafkaStreams1, storeQueryParam2).get(key), is(nullValue()));
```

(depending on which branch) to be thrown, i.e. see
```
org.apache.kafka.streams.errors.InvalidStateStorePartitionException: The specified partition 1 for store source-table does not exist.

	at org.apache.kafka.streams.state.internals.WrappingStoreProvider.stores(WrappingStoreProvider.java:63)
	at org.apache.kafka.streams.state.internals.CompositeReadOnlyKeyValueStore.get(CompositeReadOnlyKeyValueStore.java:53)
	at org.apache.kafka.streams.integration.StoreQueryIntegrationTest.lambda$shouldQuerySpecificActivePartitionStores$5(StoreQueryIntegrationTest.java:223)
	at org.apache.kafka.streams.integration.StoreQueryIntegrationTest.retryUntil(StoreQueryIntegrationTest.java:579)
	at org.apache.kafka.streams.integration.StoreQueryIntegrationTest.shouldQuerySpecificActivePartitionStores(StoreQueryIntegrationTest.java:186)
```

This happens when the stream hasn't been initialized yet. I have run the test around 12k times using Intellij's JUnit testing framework without any flaky failures. The PR also does some minor refactoring regarding moving the list of predicates into their own functions.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by Kvicii — APPROVED** [review]

**Comment by mdedetrich** [comment]
@cadonna

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 1]
```diff
@@ -535,17 +532,26 @@ public void shouldQueryStoresAfterAddingAndRemovingStreamThread() throws Excepti
...
+    private Matcher<String> retrievableException() {
+        return is(
+                anyOf(
+                        containsString("Cannot get state store source-table because the stream thread is PARTITIONS_ASSIGNED, not RUNNING"),
+                        containsString("The state store, source-table, may have migrated to another instance"),
+                        containsString("Cannot get state store source-table because the stream thread is STARTING, not RUNNING"),
+                        containsString("The specified partition 1 for store source-table does not exist.")
+                )
+        );
+    }
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: Fix for indentation and typo
```suggestion
    private Matcher<String> retriableException() {
        return is(
            anyOf(
                containsString("Cannot get state store source-table because the stream thread is PARTITIONS_ASSIGNED, not RUNNING"),
                containsString("The state store, source-table, may have migrated to another instance"),
                containsString("Cannot get state store source-table because the stream thread is STARTING, not RUNNING"),
                containsString("The specified partition 1 for store source-table does not exist.")
            )
        );
    }
```
Could you please fix the typo in the whole class. It should be `retriable`. The typo was mostly existing before this PR?
Could you please also fix the indentation in the whole class? We use 4 spaces.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 2]
```diff
@@ -210,39 +213,33 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
...
+            // Assert that key is not served when wrong specific partition is requested
+            // If kafkaStreams1 is active for keyPartition, kafkaStreams2 would be active for keyDontBelongPartition
+            // So, in that case, store3 would be null and the store4 would not return the value for key as wrong partition was requested
+            if (kafkaStreams1IsActive) {
+                assertThat(store1.get(key), is(notNullValue()));
+                assertThat(getStore(kafkaStreams2, storeQueryParam2).get(key), is(nullValue()));
+                final InvalidStateStoreException exception =
                         assertThrows(InvalidStateStoreException.class, () -> getStore(kafkaStreams1, storeQueryParam2).get(key));
-                    assertThat(
+
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]:
```suggestion
```

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 3]
```diff
@@ -210,39 +213,33 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
...
+                assertThat(
                         exception.getMessage(),
                         containsString("The specified partition 1 for store source-table does not exist.")
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]:
```suggestion
                assertThat(
                    exception.getMessage(),
                    containsString("The specified partition 1 for store source-table does not exist.")
```

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 4]
```diff
@@ -210,39 +213,33 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
...
-                }
-                return true;
-            } catch (final InvalidStateStoreException exception) {
-                verifyRetrievableException(exception);
-                LOG.info("Either streams wasn't running or a re-balancing took place. Will try again.");
-                return false;
+                );
             }
-        });
+            return true;
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: Why do we still need to return something? With your change we would retry until we do not encounter any exception anymore. 
Maybe it would be better to extend `retryOnExceptionWithTimeout()` (see https://github.com/apache/kafka/blob/e4b3a3cdeb295fdd4c4434ec1a7ee77b66553ae0/clients/src/test/java/org/apache/kafka/test/TestUtils.java#L361) to include the exception for which to retry.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 5]
```diff
@@ -535,17 +532,26 @@ public void shouldQueryStoresAfterAddingAndRemovingStreamThread() throws Excepti
         });
     }
 
+    private Matcher<String> retrievableException() {
+        return is(
+                anyOf(
+                        containsString("Cannot get state store source-table because the stream thread is PARTITIONS_ASSIGNED, not RUNNING"),
+                        containsString("The state store, source-table, may have migrated to another instance"),
+                        containsString("Cannot get state store source-table because the stream thread is STARTING, not RUNNING"),
+                        containsString("The specified partition 1 for store source-table does not exist.")
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: I see that this might happen, but I am wondering if we can solve that differently. The reason is that a bug could also throw that exception and we would just retry. Due to exceeding the timeout we would probably understand that it is a bug, but it seems really hard to investigate.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 6]
```diff
@@ -210,39 +213,33 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
...
-                if (kafkaStreams1IsActive) {
-                    assertThat(store1.get(key), is(notNullValue()));
-                    assertThat(getStore(kafkaStreams2, storeQueryParam2).get(key), is(nullValue()));
-                    final InvalidStateStoreException exception =
+                    StoreQueryParameters.<ReadOnlyKeyValueStore<Integer, Integer>>fromNameAndType(TABLE_NAME, keyValueStore())
+                            .withPartition(keyDontBelongPartition);
+            // Assert that key is not served when wrong specific partition is requested
+            // If kafkaStreams1 is active for keyPartition, kafkaStreams2 would be active for keyDontBelongPartition
+            // So, in that case, store3 would be null and the store4 would not return the value for key as wrong partition was requested
+            if (kafkaStreams1IsActive) {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: I am wondering why we have three times the same if-statement `if (kafkaStreams1IsActive)`. Can we not merge the three if-statements together?

**Review by cadonna — COMMENTED** [review]
@mdedetrich Thank you for the PR!

Here my feedback!

I have to admit that this test is really complex since rebalances can happen at any time.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java (hunk 5 above)
Comment (mdedetrich) [reply streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: Agreed but if this is the case (and its a consistent bug) it would hit the timeout. Tbh investigating this was a bit out of my depth, maybe its wise to do a separate PR if we can find a better solution?

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java (hunk 6 above)
Comment (mdedetrich) [reply streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: We could, I have a suspicion that its done this way so the test is easier to read since the flow is more linear but I am not against merging the statements if you think its better.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java (hunk 4 above)
Comment (mdedetrich) [reply streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: So the original reason for the return value is so that you can easily replace it with `until`. Regarding extending the `retryOnExceptionWithTimeout()`, the reason why `retryUntil` is handcoded is because its keeping the same unwrapping `RuntimeException` that the original `until` did (see https://github.com/apache/kafka/blob/4c9eeef5b2dff9a4f0977fbc5ac7eaaf930d0d0e/streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java#L559-L563).

I don't know the original reason why it was done this way but the intention was to minimize the amount of changes as to not increase the flakiness (but this is a hunch). Are you happy with just removing the `return` value for now or do you also want to extend ``retryOnExceptionWithTimeout()`?

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java (hunk 1 above)
Comment (mdedetrich) [reply streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: Done

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java (hunk 6 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: We cannot merge the first two if-statements because in the first if-statement `store1` and `store2` are assigned and both variables are used in each of the branches in the second if-statement.
AFAIS, we could merge the second and the third if-statement. I leave it up to you whether the two if-statements should be merged or not. I am fine either way, with a slight preference to merge them. But I also understand that this is not really needed in this PR.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 7]
```diff
@@ -180,7 +183,7 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
 
         // Assert that all messages in the first batch were processed in a timely manner
         assertThat(semaphore.tryAcquire(batch1NumMessages, 60, TimeUnit.SECONDS), is(equalTo(true)));
-        until(() -> {
+        retryUntil(() -> {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: In the PR description you state:

> The first reason is that the test uses the until method which just fails whenever a retryable error occurs.

I doubt this statement, because within the `until()` the code verifies for retrieable exceptions and returns `false` if it encounters a retrieable exception which would trigger a further iteration over the code passed into the `until()`.
You can verify this behavior by running the following test:
```
public void testUntil() throws Exception {
  until(() -> {
      try {
          throw new InvalidStateStorePartitionException(
              "The specified partition 1 for store source-table does not exist.");
      } catch (final InvalidStateStoreException exception) {
          verifyRetriableException(exception);
          LOG.info("Either streams wasn't running or a re-balancing took place. Will try again.");
          return false;
      }
  });
}
``` 
This code will retry throwing the exception until the timeout is exceeded.
You will  notice that when the timeout is exceeded, the test passes. This is definitely a bug!
The following rewrite to `until()` will fix this bug:
```
  private static void until(final TestCondition condition) {
      boolean success = false;
      final long deadline = System.currentTimeMillis() + IntegrationTestUtils.DEFAULT_TIMEOUT;
      boolean deadlineExceeded = System.currentTimeMillis() >= deadline;
      while (!success && !deadlineExceeded) {
          try {
              success = condition.conditionMet();
              Thread.sleep(500L);
          } catch (final RuntimeException e) {
              throw e;
          } catch (final Exception e) {
              throw new RuntimeException(e);
          } finally {
              deadlineExceeded = System.currentTimeMillis() >= deadline;
          }
      }
      if (deadlineExceeded) {
          fail("Test execution timed out");
      }
  }
``` 

In summary, I think method `until()` does not cause any flakiness.

If you agree with my argumentation could you remove again `retryUntil()`, add the fix to `until()`, and adapt the PR description?

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java (hunk 4 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: See my comment above about `until()`.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java (hunk 5 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: I looked again at this code and discussed it with others and I think that might indeed be one cause of the flakiness of the test. My previous concerns proofed wrong. Thank you a lot for the investigation.

**Review by cadonna — COMMENTED** [review]
@mdedetrich Thank you for the updates!

Here my feedback!

My major comment is about replacing `until()` with `retryUntil()`.

**Comment by mdedetrich** [comment]
@cadonna 

Indeed you are correct, the real cause of the flakiness was not having the extra `"The specified partition 1 for store source-table does not exist."` condition. I reverted back to using `until()` and I got 10k runs without any flaky errors.

The latest commit has `retryUntil` removed and the test is using `until` just like before. I did however keep some of the refactoring that I did along with your suggested formatting/naming changes, let me know if you also want to revert some f these changes.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 8]
```diff
@@ -210,8 +213,8 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
             }
 
             final StoreQueryParameters<ReadOnlyKeyValueStore<Integer, Integer>> storeQueryParam2 =
-                StoreQueryParameters.<ReadOnlyKeyValueStore<Integer, Integer>>fromNameAndType(TABLE_NAME, keyValueStore())
-                .withPartition(keyDontBelongPartition);
+                    StoreQueryParameters.<ReadOnlyKeyValueStore<Integer, Integer>>fromNameAndType(TABLE_NAME, keyValueStore())
+                            .withPartition(keyDontBelongPartition);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: nit: we usually indent 4 spaces.
```suggestion
                StoreQueryParameters.<ReadOnlyKeyValueStore<Integer, Integer>>fromNameAndType(TABLE_NAME, keyValueStore())
                    .withPartition(keyDontBelongPartition);
```

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 9]
```diff
@@ -221,24 +224,24 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
                     assertThat(store1.get(key), is(notNullValue()));
                     assertThat(getStore(kafkaStreams2, storeQueryParam2).get(key), is(nullValue()));
                     final InvalidStateStoreException exception =
-                        assertThrows(InvalidStateStoreException.class, () -> getStore(kafkaStreams1, storeQueryParam2).get(key));
+                            assertThrows(InvalidStateStoreException.class, () -> getStore(kafkaStreams1, storeQueryParam2).get(key));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: nit
```suggestion
                        assertThrows(InvalidStateStoreException.class, () -> getStore(kafkaStreams1, storeQueryParam2).get(key));
```

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 10]
```diff
@@ -221,24 +224,24 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
...
                     assertThat(
-                        exception.getMessage(),
-                        containsString("The specified partition 1 for store source-table does not exist.")
+                            exception.getMessage(),
+                            containsString("The specified partition 1 for store source-table does not exist.")
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: nit:
```suggestion
                        exception.getMessage(),
                        containsString("The specified partition 1 for store source-table does not exist.")
```

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 11]
```diff
@@ -221,24 +224,24 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
...
                     );
                 } else {
                     assertThat(store2.get(key), is(notNullValue()));
                     assertThat(getStore(kafkaStreams1, storeQueryParam2).get(key), is(nullValue()));
                     final InvalidStateStoreException exception =
-                        assertThrows(InvalidStateStoreException.class, () -> getStore(kafkaStreams2, storeQueryParam2).get(key));
+                            assertThrows(InvalidStateStoreException.class, () -> getStore(kafkaStreams2, storeQueryParam2).get(key));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: nit:
```suggestion
                        assertThrows(InvalidStateStoreException.class, () -> getStore(kafkaStreams2, storeQueryParam2).get(key));
```

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java [hunk 12]
```diff
@@ -221,24 +224,24 @@ public void shouldQuerySpecificActivePartitionStores() throws Exception {
...
                     assertThat(
-                        exception.getMessage(),
-                        containsString("The specified partition 1 for store source-table does not exist.")
+                            exception.getMessage(),
+                            containsString("The specified partition 1 for store source-table does not exist.")
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreQueryIntegrationTest.java]: nit:
```suggestion
                        exception.getMessage(),
                        containsString("The specified partition 1 for store source-table does not exist.")
```

**Review by cadonna — COMMENTED** [review]
@mdedetrich Thank you for your updates!

I have just a couple of formatting suggestions.

More importantly, could you please adapt the PR description and fix `until()` as I proposed?

**Comment by cadonna** [comment]
@mdedetrich The builds are red due to a checkstyle issue. Could you please fix the issue?

**Comment by mdedetrich** [comment]
@cadonna Apologies, missed the part regarding `until`, there was also some issues regarding rebasing. I have fixed the `until` method as you stated as well as the various formatting issues and I checked that locally it passes checkstyle.

PR description is also updated, there is no mention of `retryUntil`. Not sure if you also want me to update the description of the original commit?

**Review by cadonna — APPROVED** [review]
Thanks for the updates!

LGTM!

I will wait until the builds are acceptable and then merge.

**Comment by cadonna** [comment]
The timeout in the ARM build seems to be unrelated since that build does also time out a lot on other PRs recently.
